h2o-genmodel-extensions/xgboost/src/main/java/hex/genmodel/algos/xgboost/XGBoostMojoReader.java

```diff
@@ -38,6 +31,44 @@ protected void readModelData() throws IOException {
 
   @Override
   protected XGBoostMojoModel makeModel(String[] columns, String[][] domains, String responseColumn) {
-    return new XGBoostMojoModel(columns, domains, responseColumn);
+    byte[] boosterBytes;
+    try {
+      boosterBytes = readblob("boosterBytes");
+    } catch (IOException e) {
+      throw new IllegalStateException("MOJO is corrupted: cannot read the serialized Booster", e);
+    }
+    if (useJavaScoring()) {
+      return makeJavaModel(columns, domains, responseColumn, boosterBytes);
+    } else {
+      return makeNativeModel(columns, domains, responseColumn, boosterBytes);
+    }
+  }
+
+  public static boolean useJavaScoring() {
+    return Boolean.getBoolean("sys.ai.h2o.xgboost.scoring.java.enable");
   }
+
+  private  XGBoostMojoModel makeNativeModel(String[] columns, String[][] domains, String responseColumn,
+                                            byte[] boosterBytes) {
+    XGBoostNativeMojoModel model = new XGBoostNativeMojoModel(columns, domains, responseColumn);
+    InputStream is = new ByteArrayInputStream(boosterBytes);
+    try {
+      model._booster = BoosterHelper.loadModel(is);
+    } catch (Exception xgBoostError) {
+      throw new IllegalStateException("Unable to load XGBooster", xgBoostError);
+    }
+    return model;
+  }
+
+  private  XGBoostMojoModel makeJavaModel(String[] columns, String[][] domains, String responseColumn,
+                                          byte[] boosterBytes) {
+    XGBoostJavaMojoModel model = new XGBoostJavaMojoModel(columns, domains, responseColumn);
+    try (InputStream is = new ByteArrayInputStream(boosterBytes)) {
+      model._predictor = new Predictor(is);
```

Comment:
Shouldn't `XGBoostJavaMojoModel` construct it's own `Predictor` ? Maybe we could pass the byte stream in constructor and have the predictor created right away, instead of doing it manually afterwards. I'd try to create fully-functional objects by means of constructor.

Reply:
@Pscheidl good thinking, will change, thx